### PR TITLE
Various small API improvements of Cluster Sharding

### PR DESCRIPTION
**Pull request by patriknw**
Nothing advanced or difficult. Several small issues touching cluster sharding. See individual commits for details.

Issues:
- #17364
- #17158
- #15330
- #15619
- #15614
- #16422

### CODE REVIEW

**Comment by patriknw** [comment]
this failure was wrong javadoc

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala [hunk 1]
```diff
@@ -51,40 +54,40 @@ import scala.util.Success
  * several nodes in the cluster and you want to be able to interact with them using their
  * logical identifier, but without having to care about their physical location in the cluster,
  * which might also change over time. It could for example be actors representing Aggregate Roots in
- * Domain-Driven Design terminology. Here we call these actors "entries". These actors
+ * Domain-Driven Design terminology. Here we call these actors "entities". These actors
```
Comment (ktoso) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala]: very good rename

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala [hunk 2]
```diff
@@ -175,88 +179,59 @@ class ClusterSharding(system: ExtendedActorSystem) extends Extension {
...
-   * @param typeName the name of the entry type
-   * @param entryProps the `Props` of the entry actors that will be created by the `ShardRegion`,
-   *   if not defined (None) the `ShardRegion` on this node will run in proxy only mode, i.e.
-   *   it will delegate messages to other `ShardRegion` actors on other nodes, but not host any
-   *   entry actors itself
-   * @param roleOverride specifies that this entry type requires cluster nodes with a specific role.
-   *   if not defined (None), then defaults to standard behavior of using Role (if any) from configuration
-   * @param rememberEntries true if entry actors shall created be automatically restarted upon `Shard`
-   *   restart. i.e. if the `Shard` is started on a different `ShardRegion` due to rebalance or crash.
-   * @param idExtractor partial function to extract the entry id and the message to send to the
```
Comment (ktoso) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala]: `extractEntityId` is now not documented, accidental omission?

**Comment by ktoso** [comment]
Minor comments in line, LGTM!

**Comment by patriknw** [comment]
Thanks for the review, @ktoso 
I will fix the minor things, and merge this first in my queue of pull requests.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala (hunk 2 above)
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala]: thx

File: akka-cluster-sharding/src/main/resources/reference.conf [hunk 3]
```diff
@@ -14,44 +14,60 @@ akka.cluster.sharding {
...
   guardian-name = sharding
   
+  # Specifies that entities runs on cluster nodes with a specific role.
+  # If the role is not specified (or empty) all nodes in the cluster are used.
+  role = ""
+  
+  # When this is set to 'on' the active entity actors will automatically be restarted
+  # upon Shard restart. i.e. if the Shard is started on a different ShardRegion
+  # due to rebalance or crash.
+  remember-entities = off
```
Comment (anderius) [inline akka-cluster-sharding/src/main/resources/reference.conf]: Nice!

File: akka-docs/rst/scala/cluster-sharding.rst [hunk 4]
```diff
@@ -238,41 +239,42 @@ Proxy Only Mode
...
+
+Remembering Entities
+--------------------
+
+The list of entities in each ``Shard`` can be made persistent (durable) by setting
+the ``rememberEntities`` flag to true in ``ClusterShardingSettings`` when calling 
+``ClusterSharding.start``. When configured to remember entities, whenever a ``Shard`` 
+is rebalanced onto another node or recovers after a crash it will recreate all the
+entities which were previously running in that ``Shard``. To permanently stop entities, 
+a ``Passivate`` message must be sent to the parent the ``Shard``, otherwise the
```
Comment (anderius) [inline akka-docs/rst/scala/cluster-sharding.rst]: Nice documentation, very much appreciated!

PS:Not native English-speaker, but is there missing an 'to' there? "parent to the ``Shard``?

Comment (patriknw) [inline akka-docs/rst/scala/cluster-sharding.rst]: ah, `Passivate` should be sent to the parent actor, period. That actor is the `Shard` actor, but that not really something that the user should care about.

**Pull request merged**